{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7edee4e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../src/')\n",
    "\n",
    "from utils_tiramisu import *\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "# this is the same TIRAMISU_PATH as shown in start_here.ipynb\n",
    "TIRAMISU_PATH = \n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import pickle\n",
    "import gensim\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "import logging\n",
    "logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.DEBUG)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60bedcd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "\n",
    "sns.color_palette(\"Set1\")\n",
    "\n",
    "matplotlib.rc('font', family='Helvetica') \n",
    "matplotlib.rc('pdf', fonttype=42)\n",
    "matplotlib.rc('text', usetex='false') \n",
    "matplotlib.rcParams['axes.unicode_minus'] = False\n",
    "matplotlib.rcParams.update({\"axes.labelsize\": 7,\n",
    "\"xtick.labelsize\": 7,\n",
    "\"ytick.labelsize\": 7,\n",
    "\"legend.fontsize\": 5,\n",
    "\"font.size\":7})\n",
    "\n",
    "\n",
    "matplotlib.rcParams['xtick.major.size'] = 2\n",
    "matplotlib.rcParams['xtick.major.width'] = 0.5\n",
    "matplotlib.rcParams['xtick.minor.size'] = 2\n",
    "matplotlib.rcParams['xtick.minor.width'] = 0.5\n",
    "\n",
    "matplotlib.rcParams['ytick.major.size'] = 2\n",
    "matplotlib.rcParams['ytick.major.width'] = 0.5\n",
    "matplotlib.rcParams['ytick.minor.size'] = 2\n",
    "matplotlib.rcParams['ytick.minor.width'] = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e055007a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from openTSNE import TSNE\n",
    "from openTSNE import affinity, initialization, TSNEEmbedding\n",
    "from openTSNE.affinity import Affinities\n",
    "import scipy as sp\n",
    "from scipy import sparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659648aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mycallback(iteration, error, embedding):\n",
    "    Zs.append(embedding.copy())\n",
    "    kls.append(error)\n",
    "    n_iter.append(iteration)\n",
    "    \n",
    "    \n",
    "def projects(x):\n",
    "\n",
    "    if x['secondary'] == 'human sequence':\n",
    "        return \"Human Genome Project\"\n",
    "    elif x['secondary'] == \"Box026-010.pdf\" and x['folders'] == \"Large scale sequence\":\n",
    "        return \"LSAC\"\n",
    "    elif x['folders'] == 'Large scale sequence':\n",
    "        return \"\"\n",
    "    elif x['folders'] == 'sequencingrampupfiles':\n",
    "        return \"Human Genome Project\"\n",
    "    elif x['folders'] == \"eMERGE\":\n",
    "        return \"eMERGE\"\n",
    "    elif x['folders'] == \"PAGE\":\n",
    "        return \"PAGE\"\n",
    "    elif x['folders'] == \"ENCODE\":\n",
    "        return \"ENCODE\"\n",
    "    elif x['folders'] == 'modENCODE':\n",
    "        return 'modENCODE'\n",
    "    elif x['folders'] == 'ELSI':\n",
    "        return 'ELSI'\n",
    "    elif x['folders'] == 'Celera':\n",
    "        return \"Human Genome Project\"\n",
    "    elif x['folders'] == \"H3Africa\":\n",
    "        return \"H3Africa\"\n",
    "    elif x['folders'] == 'Sequence target files':\n",
    "        return \"LSAC\"\n",
    "    elif x['folders'] == \"Haplotype Map Project\":\n",
    "        return \"HapMap\"\n",
    "    elif x['folders'] == \"GWAS materials\":\n",
    "        return \"GWAS\"\n",
    "    else:\n",
    "        return x['folders']\n",
    "\n",
    "colors = {'ELSI': '#F04A3B',\n",
    " 'GWAS': '#9FB13A',\n",
    " 'HapMap': '#E1BE15',\n",
    " 'LSAC': '#51AF4D',\n",
    " 'ENCODE': '#095393',\n",
    " 'modENCODE': '#AC5D95',\n",
    " 'eMERGE': 'maroon',\n",
    " 'Human Genome Project': 'black',\n",
    " 'H3Africa': '#06B4DB',\n",
    " 'PAGE': '#4A4EA1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbd6e955",
   "metadata": {},
   "outputs": [],
   "source": [
    "together = pd.read_parquet(f'../models/stopwords_filtering/filtered_text_240320_{0.1}_removed_22070.parquet')\n",
    "\n",
    "len(together[['text', 'nodeID']].set_index('nodeID').groupby('nodeID').apply(lambda x : x.to_numpy().tolist()).to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "188d6b31",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "all_pdfs = return_from_neo4j(\"\"\"\n",
    "match (n:Folder) - [:CONTAINS] -> (e:File) - [:SPLIT_INTO] -> (c:File) - [:CONVERT_TO] -> (f:File) \n",
    "where e.fileExtension = 'pdf' and f.fileExtension = 'png' \n",
    "return c.nodeID as nodeID, e.originalPath as path, e.fileExtension as fileExtension\n",
    "\"\"\")\n",
    "\n",
    "all_ms = return_from_neo4j(\"\"\"\n",
    "match (n:Folder) - [:CONTAINS] -> (e:File) \n",
    "where e.fileExtension in ['doc', 'docx', 'ppt', 'pptx'] \n",
    "return e.nodeID as nodeID, e.originalPath as path, e.fileExtension as fileExtension\n",
    "\"\"\")\n",
    "\n",
    "\n",
    "for fraction in tqdm([0, 0.01, 0.05, 0.1, 0.2, 0.5]):\n",
    "\n",
    "    # this cached file is created in `doc2Vec_remove_stopwords.ipynb`\n",
    "    together = pd.read_parquet(f'../models/stopwords_filtering/filtered_text_240320_{fraction}_removed_22070.parquet')\n",
    "\n",
    "    to_put_into_gensim = together[['text', 'nodeID']].set_index('nodeID').groupby('nodeID').apply(lambda x : x.to_numpy().tolist()).to_dict()\n",
    "    \n",
    "    corpus_list = []\n",
    "\n",
    "    for document in tqdm(to_put_into_gensim, total = len(to_put_into_gensim)):\n",
    "        temp_corpus = \"\"\n",
    "        temp_key = \"\"\n",
    "        for i in to_put_into_gensim[document]:\n",
    "            temp_corpus = temp_corpus + i[0] + \" \"\n",
    "            temp_key = document\n",
    "        corpus_list.append(gensim.models.doc2vec.TaggedDocument([token.strip() for token in temp_corpus.split()], [temp_key]))\n",
    "        \n",
    "    model = gensim.models.doc2vec.Doc2Vec(vector_size=100, min_count=2, epochs=50, workers = 10)\n",
    "    model.build_vocab(corpus_list)\n",
    "    model.train(corpus_list, total_examples=model.corpus_count, epochs=model.epochs)\n",
    "    model.save(f'../models/stopwords_filtering/doc2vec-model-filtered-text-{fraction}-removed-on-010924-corpus_22070')\n",
    "    \n",
    "    indices = []\n",
    "    columns = []\n",
    "    similarity_list = []\n",
    "    for i in range(len(corpus_list)):\n",
    "        doc_id = corpus_list[i].tags[0]\n",
    "        columns.append(doc_id)\n",
    "        inferred_vector = model.infer_vector(corpus_list[i].words)\n",
    "        sims = model.dv.most_similar([inferred_vector], topn=len(corpus_list))\n",
    "\n",
    "\n",
    "        similarities = []\n",
    "        docids = []\n",
    "        for docid, sim in sims:\n",
    "            similarities.append(sim)\n",
    "            docids.append(docid)\n",
    "        similarity_list.append(similarities)\n",
    "        indices.append(docids)\n",
    "    dfs_to_concat = []\n",
    "\n",
    "    for i in range(len(indices)):\n",
    "        df = pd.DataFrame(similarity_list[i], columns = [columns[i]], index = indices[i])\n",
    "        dfs_to_concat.append(df)\n",
    "    maps = pd.concat(dfs_to_concat, axis = 1)\n",
    "    with open(f'../models/stopwords_filtering/map-filtered-text-{fraction}-removed-on-010924-corpus_22070.pkl', 'wb') as f:\n",
    "        pickle.dump(maps, f)\n",
    "    maps = maps.sort_index(axis = 0)\n",
    "    maps = maps.sort_index(axis = 1)\n",
    "    maps = maps.T\n",
    "    maps= maps.rename_axis('path').reset_index()\n",
    "\n",
    "    maps['all_folders'] = maps['path'].apply(lambda x: Path(x).parts)\n",
    "    maps['folders'] = maps['all_folders'].apply(lambda x: x[2])\n",
    "#     print(maps.groupby('folders').count().sort_values('path', ascending= False).index.to_list())\n",
    "    # maps = maps.loc[maps.folders.isin(maps.groupby('folders').count().sort_values('path', ascending= False).head(8).index.to_list())]\n",
    "    # maps = maps.loc[maps.folders.isin(['ELSI', 'Haplotype Map Project', 'Sequence target files', 'GWAS materials'])]\n",
    "    temp = maps.path.tolist()\n",
    "    temp.append('folders')\n",
    "    temp.append('path')\n",
    "    temp.append('all_folders')\n",
    "    maps = maps[temp]\n",
    "    folders = maps.pop('folders')\n",
    "    paths = maps.pop('path')\n",
    "    all_folders= maps.pop('all_folders')\n",
    "    maps_2 = pd.DataFrame(np.triu(maps, k = 1) + np.triu(maps).T, index = maps.index, columns = maps.columns)\n",
    "    \n",
    "    A = affinity.Uniform(\n",
    "        maps_2.to_numpy(),\n",
    "        k_neighbors=10,\n",
    "        n_jobs=-1,\n",
    "        verbose=1,\n",
    "        random_state=42,\n",
    "    )\n",
    "    sp.sparse.save_npz(f\"../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_affinities_P_22070\", A.P)\n",
    "    I = initialization.pca(maps_2.to_numpy(), random_state=42)\n",
    "    np.save(f\"../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_initialization_22070\", I)\n",
    "    Zs = []\n",
    "    kls = []\n",
    "    n_iter = []\n",
    "    E = TSNEEmbedding(I, A, n_jobs=-1, random_state=42, verbose=True)\n",
    "\n",
    "    # early exaggeration\n",
    "    E = E.optimize(n_iter=125, exaggeration=12, momentum=0.5, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=50)\n",
    "\n",
    "    # exaggeration annealing\n",
    "    exs = np.linspace(12,1,125)\n",
    "    for i in range(125):\n",
    "        if (i+1)%50 == 0:\n",
    "            E = E.optimize(n_iter=1, exaggeration=exs[i], momentum=0.8, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=1)\n",
    "\n",
    "        else:\n",
    "            E = E.optimize(n_iter=1, exaggeration=exs[i], momentum=0.8, n_jobs=-1, verbose=True)\n",
    "\n",
    "    # final optimization without exaggeration\n",
    "    E = E.optimize(n_iter=2000, exaggeration=1, momentum=0.8, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=50)\n",
    "    \n",
    "    tsne=np.array(E)\n",
    "\n",
    "    #save\n",
    "    np.save(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne_22070', tsne)\n",
    "    \n",
    "    merged = pd.merge(pd.merge(pd.DataFrame(tsne, columns = [\"dim1\", \"dim2\"]), folders.reset_index(drop = True), left_index = True, right_index = True),\n",
    "                  paths.reset_index(drop = True), left_index = True, right_index = True)\n",
    "    \n",
    "    merged['secondary'] = merged['path'].apply(lambda x: x.split('/')[3] if len(x.split('/')) > 2 else None)\n",
    "    merged['tertiary'] = merged['path'].apply(lambda x: x.split('/')[4] if len(x.split('/')) > 4 else x.split('/')[3])\n",
    "    \n",
    "    merged['projects'] = merged.apply(lambda x: projects(x), axis = 1)\n",
    "    \n",
    "    merged['color'] = merged['projects'].map(colors)\n",
    "    merged['color'] = merged['color'].fillna(\"#B8BABC\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    fig, ax = plt.subplots(1, 1, figsize = (20 , 20), dpi = 300 )\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    sns.scatterplot(data = merged.loc[merged.color == \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =50)\n",
    "    sns.scatterplot(data = merged.loc[merged.color != \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 50)\n",
    "\n",
    "    plt.legend(title='Folder structure', loc='upper right',\\\n",
    "               bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.legend().remove()\n",
    "    sns.despine()\n",
    "    plt.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    plt.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "    plt.savefig(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne_22070.png', transparent = True, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b76f8b0a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfd14c0a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fa427b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for fraction in tqdm([0, 0.01, 0.05, 0.1, 0.2, 0.5]):\n",
    "    \n",
    "    together = pd.read_parquet(f'../models/stopwords_filtering/filtered_text_240320_{fraction}_removed_22070.parquet')\n",
    "\n",
    "    to_put_into_gensim = together[['text', 'nodeID']].set_index('nodeID').groupby('nodeID').apply(lambda x : x.to_numpy().tolist()).to_dict()\n",
    "    \n",
    "    corpus_list = []\n",
    "\n",
    "    for document in tqdm(to_put_into_gensim, total = len(to_put_into_gensim)):\n",
    "        temp_corpus = \"\"\n",
    "        temp_key = \"\"\n",
    "        for i in to_put_into_gensim[document]:\n",
    "            temp_corpus = temp_corpus + i[0] + \" \"\n",
    "            temp_key = document\n",
    "        corpus_list.append(gensim.models.doc2vec.TaggedDocument([token.strip() for token in temp_corpus.split()], [temp_key]))\n",
    "        \n",
    "    model = gensim.models.doc2vec.Doc2Vec(vector_size=100, min_count=5, epochs=50, workers = 10)\n",
    "    model.build_vocab(corpus_list)\n",
    "    model.train(corpus_list, total_examples=model.corpus_count, epochs=model.epochs)\n",
    "    model.save(f'../models/stopwords_filtering/doc2vec-model-filtered-text-{fraction}-removed-on-010924-corpus-min-count-5_22070')\n",
    "    \n",
    "    indices = []\n",
    "    columns = []\n",
    "    similarity_list = []\n",
    "    for i in range(len(corpus_list)):\n",
    "        doc_id = corpus_list[i].tags[0]\n",
    "        columns.append(doc_id)\n",
    "        inferred_vector = model.infer_vector(corpus_list[i].words)\n",
    "        sims = model.dv.most_similar([inferred_vector], topn=len(corpus_list))\n",
    "\n",
    "\n",
    "        similarities = []\n",
    "        docids = []\n",
    "        for docid, sim in sims:\n",
    "            similarities.append(sim)\n",
    "            docids.append(docid)\n",
    "        similarity_list.append(similarities)\n",
    "        indices.append(docids)\n",
    "    dfs_to_concat = []\n",
    "\n",
    "    for i in range(len(indices)):\n",
    "        df = pd.DataFrame(similarity_list[i], columns = [columns[i]], index = indices[i])\n",
    "        dfs_to_concat.append(df)\n",
    "    maps = pd.concat(dfs_to_concat, axis = 1)\n",
    "    with open(f'../models/stopwords_filtering/map-filtered-text-{fraction}-removed-on-010924-corpus-min-count-5_22070.pkl', 'wb') as f:\n",
    "        pickle.dump(maps, f)\n",
    "    maps = maps.sort_index(axis = 0)\n",
    "    maps = maps.sort_index(axis = 1)\n",
    "    maps = maps.T\n",
    "    maps= maps.rename_axis('path').reset_index()\n",
    "\n",
    "    maps['all_folders'] = maps['path'].apply(lambda x: Path(x).parts)\n",
    "    maps['folders'] = maps['all_folders'].apply(lambda x: x[2])\n",
    "#     print(maps.groupby('folders').count().sort_values('path', ascending= False).index.to_list())\n",
    "    # maps = maps.loc[maps.folders.isin(maps.groupby('folders').count().sort_values('path', ascending= False).head(8).index.to_list())]\n",
    "    # maps = maps.loc[maps.folders.isin(['ELSI', 'Haplotype Map Project', 'Sequence target files', 'GWAS materials'])]\n",
    "    temp = maps.path.tolist()\n",
    "    temp.append('folders')\n",
    "    temp.append('path')\n",
    "    temp.append('all_folders')\n",
    "    maps = maps[temp]\n",
    "    folders = maps.pop('folders')\n",
    "    paths = maps.pop('path')\n",
    "    all_folders= maps.pop('all_folders')\n",
    "    maps_2 = pd.DataFrame(np.triu(maps, k = 1) + np.triu(maps).T, index = maps.index, columns = maps.columns)\n",
    "    \n",
    "    A = affinity.Uniform(\n",
    "        maps_2.to_numpy(),\n",
    "        k_neighbors=10,\n",
    "        n_jobs=-1,\n",
    "        verbose=1,\n",
    "        random_state=42,\n",
    "    )\n",
    "    sp.sparse.save_npz(f\"../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_affinities_P-min-count-5_22070\", A.P)\n",
    "    I = initialization.pca(maps_2.to_numpy(), random_state=42)\n",
    "    np.save(f\"../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_initialization-min-count-5_22070\", I)\n",
    "    Zs = []\n",
    "    kls = []\n",
    "    n_iter = []\n",
    "    E = TSNEEmbedding(I, A, n_jobs=-1, random_state=42, verbose=True)\n",
    "\n",
    "    # early exaggeration\n",
    "    E = E.optimize(n_iter=125, exaggeration=12, momentum=0.5, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=50)\n",
    "\n",
    "    # exaggeration annealing\n",
    "    exs = np.linspace(12,1,125)\n",
    "    for i in range(125):\n",
    "        if (i+1)%50 == 0:\n",
    "            E = E.optimize(n_iter=1, exaggeration=exs[i], momentum=0.8, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=1)\n",
    "\n",
    "        else:\n",
    "            E = E.optimize(n_iter=1, exaggeration=exs[i], momentum=0.8, n_jobs=-1, verbose=True)\n",
    "\n",
    "    # final optimization without exaggeration\n",
    "    E = E.optimize(n_iter=2000, exaggeration=1, momentum=0.8, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=50)\n",
    "    \n",
    "    tsne=np.array(E)\n",
    "\n",
    "    #save\n",
    "    np.save(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne-min-count-5_22070', tsne)\n",
    "    \n",
    "    merged = pd.merge(pd.merge(pd.DataFrame(tsne, columns = [\"dim1\", \"dim2\"]), folders.reset_index(drop = True), left_index = True, right_index = True),\n",
    "                  paths.reset_index(drop = True), left_index = True, right_index = True)\n",
    "    \n",
    "    merged['secondary'] = merged['path'].apply(lambda x: x.split('/')[3] if len(x.split('/')) > 2 else None)\n",
    "    merged['tertiary'] = merged['path'].apply(lambda x: x.split('/')[4] if len(x.split('/')) > 4 else x.split('/')[3])\n",
    "    \n",
    "    merged['projects'] = merged.apply(lambda x: projects(x), axis = 1)\n",
    "    \n",
    "    merged['color'] = merged['projects'].map(colors)\n",
    "    merged['color'] = merged['color'].fillna(\"#B8BABC\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    fig, ax = plt.subplots(1, 1, figsize = (20 , 20), dpi = 300 )\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    sns.scatterplot(data = merged.loc[merged.color == \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =50)\n",
    "    sns.scatterplot(data = merged.loc[merged.color != \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 50)\n",
    "\n",
    "    plt.legend(title='Folder structure', loc='upper right',\\\n",
    "               bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.legend().remove()\n",
    "    sns.despine()\n",
    "    plt.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    plt.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "    plt.savefig(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne-min-count-5_22070.png', transparent = True, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "818ec79c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for fraction in tqdm([0.5]):\n",
    "\n",
    "    together = pd.read_parquet(f'../models/stopwords_filtering/filtered_text_240320_{fraction}_removed_22070.parquet')\n",
    "\n",
    "    to_put_into_gensim = together[['text', 'nodeID']].set_index('nodeID').groupby('nodeID').apply(lambda x : x.to_numpy().tolist()).to_dict()\n",
    "    \n",
    "    corpus_list = []\n",
    "\n",
    "    for document in tqdm(to_put_into_gensim, total = len(to_put_into_gensim)):\n",
    "        temp_corpus = \"\"\n",
    "        temp_key = \"\"\n",
    "        for i in to_put_into_gensim[document]:\n",
    "            temp_corpus = temp_corpus + i[0] + \" \"\n",
    "            temp_key = document\n",
    "        corpus_list.append(gensim.models.doc2vec.TaggedDocument([token.strip() for token in temp_corpus.split()], [temp_key]))\n",
    "        \n",
    "#     model = gensim.models.doc2vec.Doc2Vec(vector_size=100, min_count=5, epochs=50, workers = 10)\n",
    "#     model.build_vocab(corpus_list)\n",
    "#     model.train(corpus_list, total_examples=model.corpus_count, epochs=model.epochs)\n",
    "#     model.save(f'../models/stopwords_filtering/doc2vec-model-filtered-text-{fraction}-removed-on-010924-corpus-min-count-5_22070')\n",
    "    model = gensim.models.doc2vec.Doc2Vec.load(f'../models/stopwords_filtering/doc2vec-model-filtered-text-{fraction}-removed-on-010924-corpus-min-count-5_22070')\n",
    "    \n",
    "    indices = []\n",
    "    columns = []\n",
    "    similarity_list = []\n",
    "    for i in range(len(corpus_list)):\n",
    "        doc_id = corpus_list[i].tags[0]\n",
    "        columns.append(doc_id)\n",
    "        inferred_vector = model.infer_vector(corpus_list[i].words)\n",
    "        sims = model.dv.most_similar([inferred_vector], topn=len(corpus_list))\n",
    "\n",
    "\n",
    "        similarities = []\n",
    "        docids = []\n",
    "        for docid, sim in sims:\n",
    "            similarities.append(sim)\n",
    "            docids.append(docid)\n",
    "        similarity_list.append(similarities)\n",
    "        indices.append(docids)\n",
    "    dfs_to_concat = []\n",
    "\n",
    "    for i in range(len(indices)):\n",
    "        df = pd.DataFrame(similarity_list[i], columns = [columns[i]], index = indices[i])\n",
    "        dfs_to_concat.append(df)\n",
    "    maps = pd.concat(dfs_to_concat, axis = 1)\n",
    "    with open(f'../models/stopwords_filtering/map-filtered-text-{fraction}-removed-on-010924-corpus-min-count-5_22070.pkl', 'wb') as f:\n",
    "        pickle.dump(maps, f)\n",
    "    maps = maps.sort_index(axis = 0)\n",
    "    maps = maps.sort_index(axis = 1)\n",
    "    maps = maps.T\n",
    "    maps= maps.rename_axis('path').reset_index()\n",
    "\n",
    "    maps['all_folders'] = maps['path'].apply(lambda x: Path(x).parts)\n",
    "    maps['folders'] = maps['all_folders'].apply(lambda x: x[2])\n",
    "#     print(maps.groupby('folders').count().sort_values('path', ascending= False).index.to_list())\n",
    "    # maps = maps.loc[maps.folders.isin(maps.groupby('folders').count().sort_values('path', ascending= False).head(8).index.to_list())]\n",
    "    # maps = maps.loc[maps.folders.isin(['ELSI', 'Haplotype Map Project', 'Sequence target files', 'GWAS materials'])]\n",
    "    temp = maps.path.tolist()\n",
    "    temp.append('folders')\n",
    "    temp.append('path')\n",
    "    temp.append('all_folders')\n",
    "    maps = maps[temp]\n",
    "    folders = maps.pop('folders')\n",
    "    paths = maps.pop('path')\n",
    "    all_folders= maps.pop('all_folders')\n",
    "    maps_2 = pd.DataFrame(np.triu(maps, k = 1) + np.triu(maps).T, index = maps.index, columns = maps.columns)\n",
    "    \n",
    "    A = affinity.Uniform(\n",
    "        maps_2.to_numpy(),\n",
    "        k_neighbors=10,\n",
    "        n_jobs=-1,\n",
    "        verbose=1,\n",
    "        random_state=42,\n",
    "    )\n",
    "    sp.sparse.save_npz(f\"../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_affinities_P-min-count-5_22070\", A.P)\n",
    "    I = initialization.pca(maps_2.to_numpy(), random_state=42)\n",
    "    np.save(f\"../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_initialization-min-count-5_22070\", I)\n",
    "    Zs = []\n",
    "    kls = []\n",
    "    n_iter = []\n",
    "    E = TSNEEmbedding(I, A, n_jobs=-1, random_state=42, verbose=True)\n",
    "\n",
    "    # early exaggeration\n",
    "    E = E.optimize(n_iter=125, exaggeration=12, momentum=0.5, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=50)\n",
    "\n",
    "    # exaggeration annealing\n",
    "    exs = np.linspace(12,1,125)\n",
    "    for i in range(125):\n",
    "        if (i+1)%50 == 0:\n",
    "            E = E.optimize(n_iter=1, exaggeration=exs[i], momentum=0.8, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=1)\n",
    "\n",
    "        else:\n",
    "            E = E.optimize(n_iter=1, exaggeration=exs[i], momentum=0.8, n_jobs=-1, verbose=True)\n",
    "\n",
    "    # final optimization without exaggeration\n",
    "    E = E.optimize(n_iter=2000, exaggeration=1, momentum=0.8, n_jobs=-1, verbose=True, callbacks=mycallback, callbacks_every_iters=50)\n",
    "    \n",
    "    tsne=np.array(E)\n",
    "\n",
    "    #save\n",
    "    np.save(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne-min-count-5_22070', tsne)\n",
    "    \n",
    "    merged = pd.merge(pd.merge(pd.DataFrame(tsne, columns = [\"dim1\", \"dim2\"]), folders.reset_index(drop = True), left_index = True, right_index = True),\n",
    "                  paths.reset_index(drop = True), left_index = True, right_index = True)\n",
    "    \n",
    "    merged['secondary'] = merged['path'].apply(lambda x: x.split('/')[3].split(\"---\")[0] if len(x.split('/')) > 2 else None)\n",
    "    merged['tertiary'] = merged['path'].apply(lambda x: x.split('/')[4].split(\"---\")[0] if len(x.split('/')) > 4 else x.split('/')[3].split('---')[0])\n",
    "    \n",
    "    merged['projects'] = merged.apply(lambda x: projects(x), axis = 1)\n",
    "    \n",
    "    merged['color'] = merged['projects'].map(colors)\n",
    "    merged['color'] = merged['color'].fillna(\"#B8BABC\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    fig, ax = plt.subplots(1, 1, figsize = (20 , 20), dpi = 300 )\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    sns.scatterplot(data = merged.loc[merged.color == \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =50)\n",
    "    sns.scatterplot(data = merged.loc[merged.color != \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 50)\n",
    "\n",
    "    plt.legend(title='Folder structure', loc='upper right',\\\n",
    "               bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.legend().remove()\n",
    "    sns.despine()\n",
    "    plt.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    plt.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "    plt.savefig(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne-min-count-5_22070.png', transparent = True, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adea293d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tsne(fraction, ax):\n",
    "    with open(f'../models/stopwords_filtering/map-filtered-text-{fraction}-removed-on-010924-corpus_22070.pkl', 'rb') as f:\n",
    "        maps = pickle.load(f)\n",
    "    maps = maps.sort_index(axis = 0)\n",
    "    maps = maps.sort_index(axis = 1)\n",
    "    maps = maps.T\n",
    "    maps= maps.rename_axis('path').reset_index()\n",
    "\n",
    "    maps['all_folders'] = maps['path'].apply(lambda x: Path(x).parts)\n",
    "    maps['folders'] = maps['all_folders'].apply(lambda x: x[2])\n",
    "#     print(maps.groupby('folders').count().sort_values('path', ascending= False).index.to_list())\n",
    "    # maps = maps.loc[maps.folders.isin(maps.groupby('folders').count().sort_values('path', ascending= False).head(8).index.to_list())]\n",
    "    # maps = maps.loc[maps.folders.isin(['ELSI', 'Haplotype Map Project', 'Sequence target files', 'GWAS materials'])]\n",
    "    temp = maps.path.tolist()\n",
    "    temp.append('folders')\n",
    "    temp.append('path')\n",
    "    temp.append('all_folders')\n",
    "    maps = maps[temp]\n",
    "    folders = maps.pop('folders')\n",
    "    paths = maps.pop('path')\n",
    "    all_folders= maps.pop('all_folders')\n",
    "    \n",
    "    \n",
    "    tsne = np.load(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne_22070.npy')\n",
    "    merged = pd.merge(pd.merge(pd.DataFrame(tsne, columns = [\"dim1\", \"dim2\"]), folders.reset_index(drop = True), left_index = True, right_index = True),\n",
    "                  paths.reset_index(drop = True), left_index = True, right_index = True)\n",
    "    \n",
    "    merged['secondary'] = merged['path'].apply(lambda x: x.split('/')[3].split(\"---\")[0] if len(x.split('/')) > 2 else None)\n",
    "    merged['tertiary'] = merged['path'].apply(lambda x: x.split('/')[4].split(\"---\")[0] if len(x.split('/')) > 4 else x.split('/')[3].split('---')[0])\n",
    "    \n",
    "    merged['projects'] = merged.apply(lambda x: projects(x), axis = 1)\n",
    "    \n",
    "    merged['color'] = merged['projects'].map(colors)\n",
    "    merged['color'] = merged['color'].fillna(\"#B8BABC\")\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    sns.scatterplot(data = merged.loc[merged.color == \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =50)\n",
    "    sns.scatterplot(data = merged.loc[merged.color != \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 50)\n",
    "\n",
    "#     plt.legend(title='Folder structure', loc='upper right',\\\n",
    "#                bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.legend().remove()\n",
    "    sns.despine()\n",
    "\n",
    "    \n",
    "    ax.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    ax.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4554a46f",
   "metadata": {},
   "outputs": [],
   "source": [
    "logging.basicConfig(level=logging.ERROR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38eb062",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time \n",
    "layout = [\n",
    "    [0, 1, 2],\n",
    "    [3, 4, 5],\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplot_mosaic(layout, figsize=(35,35))\n",
    "fractions = [0, 0.01, 0.05, 0.1, 0.2, 0.5]\n",
    "for i in range(len(fractions)):\n",
    "    plot_tsne(fractions[i], axes[i])\n",
    "    axes[i].set_title(f\"{int(fractions[i] * 100)}% of tokens removed\", fontsize = 40)\n",
    "    axes[i].set_aspect('equal')\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78f4b17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tsne(fraction, ax):\n",
    "    with open(f'../models/stopwords_filtering/map-filtered-text-{fraction}-removed-on-010924-corpus-min-count-5_22070.pkl', 'rb') as f:\n",
    "        maps = pickle.load(f)\n",
    "    maps = maps.sort_index(axis = 0)\n",
    "    maps = maps.sort_index(axis = 1)\n",
    "    maps = maps.T\n",
    "    maps= maps.rename_axis('path').reset_index()\n",
    "\n",
    "    maps['all_folders'] = maps['path'].apply(lambda x: Path(x).parts)\n",
    "    maps['folders'] = maps['all_folders'].apply(lambda x: x[2])\n",
    "#     print(maps.groupby('folders').count().sort_values('path', ascending= False).index.to_list())\n",
    "    # maps = maps.loc[maps.folders.isin(maps.groupby('folders').count().sort_values('path', ascending= False).head(8).index.to_list())]\n",
    "    # maps = maps.loc[maps.folders.isin(['ELSI', 'Haplotype Map Project', 'Sequence target files', 'GWAS materials'])]\n",
    "    temp = maps.path.tolist()\n",
    "    temp.append('folders')\n",
    "    temp.append('path')\n",
    "    temp.append('all_folders')\n",
    "    maps = maps[temp]\n",
    "    folders = maps.pop('folders')\n",
    "    paths = maps.pop('path')\n",
    "    all_folders= maps.pop('all_folders')\n",
    "    \n",
    "    \n",
    "    tsne = np.load(f'../models/stopwords_filtering/filtered-text-{fraction}-removed-on-010924-corpus_tsne-min-count-5_22070.npy')\n",
    "    merged = pd.merge(pd.merge(pd.DataFrame(tsne, columns = [\"dim1\", \"dim2\"]), folders.reset_index(drop = True), left_index = True, right_index = True),\n",
    "                  paths.reset_index(drop = True), left_index = True, right_index = True)\n",
    "    \n",
    "    merged['secondary'] = merged['path'].apply(lambda x: x.split('/')[3].split(\"---\")[0] if len(x.split('/')) > 2 else None)\n",
    "    merged['tertiary'] = merged['path'].apply(lambda x: x.split('/')[4].split(\"---\")[0] if len(x.split('/')) > 4 else x.split('/')[3].split('---')[0])\n",
    "    \n",
    "    merged['projects'] = merged.apply(lambda x: projects(x), axis = 1)\n",
    "    \n",
    "    merged['color'] = merged['projects'].map(colors)\n",
    "    merged['color'] = merged['color'].fillna(\"#B8BABC\")\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    sns.scatterplot(data = merged.loc[merged.color == \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =50)\n",
    "    sns.scatterplot(data = merged.loc[merged.color != \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 50)\n",
    "\n",
    "#     plt.legend(title='Folder structure', loc='upper right',\\\n",
    "#                bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.legend().remove()\n",
    "    sns.despine()\n",
    "\n",
    "    \n",
    "    ax.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    ax.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9b3613",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time \n",
    "layout = [\n",
    "    [0, 1, 2],\n",
    "    [3, 4, 5],\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplot_mosaic(layout, figsize=(35,35))\n",
    "fractions = [0, 0.01, 0.05, 0.1, 0.2, 0.5]\n",
    "for i in range(len(fractions)):\n",
    "    plot_tsne(fractions[i], axes[i])\n",
    "    axes[i].set_title(f\"{int(fractions[i] * 100)}% of tokens removed\", fontsize = 40)\n",
    "    axes[i].set_aspect('equal')\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a27775d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c09bcf20",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'../models/stopwords_filtering/map-filtered-text-{0}-removed-on-010924-corpus_22070.pkl', 'rb') as f:\n",
    "    maps = pickle.load(f)\n",
    "maps = maps.sort_index(axis = 0)\n",
    "maps = maps.sort_index(axis = 1)\n",
    "maps = maps.T\n",
    "maps= maps.rename_axis('path').reset_index()\n",
    "\n",
    "maps['all_folders'] = maps['path'].apply(lambda x: Path(x).parts)\n",
    "maps['folders'] = maps['all_folders'].apply(lambda x: x[2])\n",
    "#     print(maps.groupby('folders').count().sort_values('path', ascending= False).index.to_list())\n",
    "# maps = maps.loc[maps.folders.isin(maps.groupby('folders').count().sort_values('path', ascending= False).head(8).index.to_list())]\n",
    "# maps = maps.loc[maps.folders.isin(['ELSI', 'Haplotype Map Project', 'Sequence target files', 'GWAS materials'])]\n",
    "temp = maps.path.tolist()\n",
    "temp.append('folders')\n",
    "temp.append('path')\n",
    "temp.append('all_folders')\n",
    "maps = maps[temp]\n",
    "folders = maps.pop('folders')\n",
    "paths = maps.pop('path')\n",
    "all_folders= maps.pop('all_folders')\n",
    "\n",
    "\n",
    "tsne = np.load(f'../models/stopwords_filtering/filtered-text-{0}-removed-on-010924-corpus_tsne_22070.npy')\n",
    "merged = pd.merge(pd.merge(pd.DataFrame(tsne, columns = [\"dim1\", \"dim2\"]), folders.reset_index(drop = True), left_index = True, right_index = True),\n",
    "              paths.reset_index(drop = True), left_index = True, right_index = True)\n",
    "\n",
    "merged['secondary'] = merged['path'].apply(lambda x: x.split('/')[3].split(\"---\")[0] if len(x.split('/')) > 2 else None)\n",
    "merged['tertiary'] = merged['path'].apply(lambda x: x.split('/')[4].split(\"---\")[0] if len(x.split('/')) > 4 else x.split('/')[3].split('---')[0])\n",
    "    \n",
    "merged['projects'] = merged.apply(lambda x: projects(x), axis = 1)\n",
    "\n",
    "merged['color'] = merged['projects'].map(colors)\n",
    "merged['color'] = merged['color'].fillna(\"#B8BABC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da1e4d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a4cba78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "sns.set_style('white', rc={\n",
    "    'xtick.bottom': True,\n",
    "    'ytick.left': True,\n",
    "})\n",
    "\n",
    "sns.color_palette(\"Set1\")\n",
    "\n",
    "matplotlib.rc('font', family='Helvetica') \n",
    "matplotlib.rc('pdf', fonttype=42)\n",
    "matplotlib.rc('text', usetex='false') \n",
    "matplotlib.rcParams['axes.unicode_minus'] = False\n",
    "matplotlib.rcParams.update({\"axes.labelsize\": 7,\n",
    "\"xtick.labelsize\": 7,\n",
    "\"ytick.labelsize\": 7,\n",
    "\"legend.fontsize\": 5,\n",
    "\"font.size\":7})\n",
    "\n",
    "\n",
    "matplotlib.rcParams['xtick.major.size'] = 2\n",
    "matplotlib.rcParams['xtick.major.width'] = 0.5\n",
    "matplotlib.rcParams['xtick.minor.size'] = 2\n",
    "matplotlib.rcParams['xtick.minor.width'] = 0.5\n",
    "\n",
    "matplotlib.rcParams['ytick.major.size'] = 2\n",
    "matplotlib.rcParams['ytick.major.width'] = 0.5\n",
    "matplotlib.rcParams['ytick.minor.size'] = 2\n",
    "matplotlib.rcParams['ytick.minor.width'] = 0.5\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize = (20 , 20), dpi = 300 )\n",
    "ax.spines['left'].set_linewidth(0)\n",
    "ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "ax.set_xticklabels([])\n",
    "ax.set_yticklabels([])\n",
    "# sns.scatterplot(data = merged.loc[merged.folders_x == \"#D3D3D3\"], x = 'dim1', y = 'dim2',\\\n",
    "#                 hue = 'folders_y', palette = merged.loc[merged.folders_x == \"#D3D3D3\"].set_index('folders_y')['folders_x'].to_dict(),\\\n",
    "#                 alpha = 0.5, ax = ax)\n",
    "# ax.get_legend().remove()\n",
    "sns.scatterplot(data = merged.loc[merged.color == \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                alpha = 1, ax = ax, linewidth =0.5, size =50)\n",
    "sns.scatterplot(data = merged.loc[merged.color != \"#B8BABC\"], x = 'dim1', y = 'dim2', \\\n",
    "                hue = 'projects', palette = merged.set_index('projects')['color'].to_dict(),\\\n",
    "                alpha = 1, ax = ax, linewidth = 0.5, size = 50)\n",
    "\n",
    "plt.legend(title='Folder structure', loc='upper right',\\\n",
    "           bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "ax.set_ylabel(None)\n",
    "ax.set_xlabel(None)\n",
    "ax.legend().remove()\n",
    "sns.despine()\n",
    "plt.tick_params(\n",
    "    axis='x',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    bottom=False,      # ticks along the bottom edge are off\n",
    "    top=False,         # ticks along the top edge are off\n",
    "    labelbottom=False) # labels along the bottom edge are off\n",
    "plt.tick_params(\n",
    "    axis='y',          # changes apply to the x-axis\n",
    "    which='both',      # both major and minor ticks are affected\n",
    "    left=False,      # ticks along the bottom edge are off\n",
    "    right=False)\n",
    "\n",
    "plt.savefig('../cache/tsne_map_legend_240320.png', transparent = True, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ad7fa8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d717ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46b4cc4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_folders = [\"1000 Genomes\", \"American Gene-Environment Study (AGES)\", \n",
    "                  \"Human Genome Project\", \"ELSI\", \"ENCODE\", \"FlyBase Materials\",\n",
    "                  \"GAIN files for Jim\", \"GGR\", \"GWAS\", \"GenPhen\", \"Genes & Environment\",\n",
    "                  \"H3Africa\", \"HapMap\", \"Healthy People\", \"PAGE\", \"ZLSAC\", \"eMERGE\",\n",
    "                  \"modENCODE\"]\n",
    "len(project_folders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b1bbfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create layout\n",
    "layout = [\n",
    "    [0, 1, 2],\n",
    "    [3, 4, 5],\n",
    "    [6, 7, 8],\n",
    "    [9, 10, 11],\n",
    "    [12, 13, 14],\n",
    "    [15, 16, 17],\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplot_mosaic(layout, figsize=(50,50))\n",
    "\n",
    "for ax, project in zip(axes.values(), sorted(project_folders, key= str.casefold)):\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    \n",
    "    if project == \"Human Genome Project\":\n",
    "        sns.scatterplot(data = merged.loc[merged.color != \"black\"], x = 'dim1', y = 'dim2', \\\n",
    "                       color = \"#d3d3d3\",\\\n",
    "                        alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged.loc[merged.color == \"black\"], x = 'dim1', y = 'dim2', \\\n",
    "                     color = \"red\",\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "        size = merged.loc[merged.color == \"black\"]\n",
    "        ax.set_title(f\"Human Genome Project\\nN={size.shape[0]}\", fontsize = 40)\n",
    "    elif project == \"ZLSAC\":\n",
    "        sns.scatterplot(data = merged.loc[merged.projects != \"LSAC\"], x = 'dim1', y = 'dim2', \\\n",
    "                         color = \"#d3d3d3\",\\\n",
    "                        alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged.loc[merged.projects == \"LSAC\"], x = 'dim1', y = 'dim2', \\\n",
    "                    color = \"red\",\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "        size = merged.loc[merged.projects == \"LSAC\"]\n",
    "        ax.set_title(f\"Sequence target files (LSAC)\\nN={size.shape[0]}\", fontsize = 40)\n",
    "    else:\n",
    "        sns.scatterplot(data = merged.loc[merged.projects != project], x = 'dim1', y = 'dim2', \\\n",
    "                         color = \"#d3d3d3\",\\\n",
    "                        alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged.loc[merged.projects == project], x = 'dim1', y = 'dim2', \\\n",
    "                    color = \"red\",\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "        if project == \"GAIN files for Jim\":\n",
    "            ax.set_title(f\"GAIN files for [redacted] (Genetic Association Information Network)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        elif project == \"GGR\":\n",
    "            ax.set_title(f\"GGR (Genomics of Gene Regulation)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize= 40)\n",
    "        elif project == \"ELSI\":\n",
    "            ax.set_title(f\"ELSI (Ethical, legal, and social implications research)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize =40)\n",
    "        elif project == \"ENCODE\":\n",
    "            ax.set_title(f\"ENCODE (The Encyclopedia of DNA Elements)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        elif project == \"modENCODE\":\n",
    "            ax.set_title(f\"modENCODE (model organisms ENCODE)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        elif project == \"GWAS\":\n",
    "            ax.set_title(f\"GWAS materials (NHGRI-EBI GWAS Catalog)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        elif project == \"HapMap\":\n",
    "            ax.set_title(f\"Haplotype Map Project (International HapMap Project)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        elif project == \"LSAC\":\n",
    "            ax.set_title(f\"Sequence target files (LSAC)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        elif project == \"PAGE\":\n",
    "            ax.set_title(f\"PAGE (Population Architecture Using Genomics and Epidemiology)\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "        else:\n",
    "            ax.set_title(f\"{project}\\nN={merged.loc[merged.projects == project].shape[0]}\", fontsize = 40)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_aspect('equal')\n",
    "    ax.legend().remove()\n",
    "    sns.despine()\n",
    "    ax.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    ax.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.savefig(\"../figures/all_project_folders_tsne_240318.pdf\", bbox_inches = \"tight\", dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b2274fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cecf8755",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_project_folders = ['1993 to 2003',\n",
    " '2003 to present',\n",
    " 'Beijing Genomics Institute (BGI)',\n",
    " 'Bioinformatics History',\n",
    " 'Bioinformatics Materials',\n",
    " 'Computational Analysis of non-coding DNA',\n",
    " 'Epigenetics Roadmap',\n",
    " 'Finishing Manual',\n",
    " 'HGP History Summer 2011',\n",
    " 'Libraries',\n",
    " 'Mouse SNP Meeting Materials',\n",
    " 'NHGRI Key early history',\n",
    " 'QA, Accuracy Discussions, Pilots MG',\n",
    " 'Yeast',\n",
    " 'comparative and organismal sequencing',\n",
    " 'files for image analysis',\n",
    " 'human genome reference consortium',\n",
    " 'model organism databases resource informatics',\n",
    "    'Celera']\n",
    "\n",
    "len(non_project_folders)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0266fa9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create layout\n",
    "layout = [\n",
    "    [0, 1, 2, 3],\n",
    "    [4, 5, 6, 7],\n",
    "    [8, 9, 10, 11],\n",
    "    [12, 13, 14, 15],\n",
    "    [16, 17, 18, '.'],\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplot_mosaic(layout, figsize=(50,50))\n",
    "\n",
    "for ax, project in zip(axes.values(), sorted(non_project_folders, key = str.lower)):\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    \n",
    "    if project == \"Celera\":\n",
    "        sns.scatterplot(data = merged.loc[merged.folders != project], x = 'dim1', y = 'dim2', \\\n",
    "                     color = \"#d3d3d3\",\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged.loc[merged.folders == project], x = 'dim1', y = 'dim2', \\\n",
    "                color = \"red\",\\\n",
    "                alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "        \n",
    "        ax.set_title(project + \"\\nN=\" + str(merged.loc[merged.folders == project].shape[0]), fontsize = 40)\n",
    "    else:\n",
    "        sns.scatterplot(data = merged.loc[merged.projects != project], x = 'dim1', y = 'dim2', \\\n",
    "                         color = \"#d3d3d3\",\\\n",
    "                        alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged.loc[merged.projects == project], x = 'dim1', y = 'dim2', \\\n",
    "                    color = \"red\",\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "        ax.set_title(project + \"\\nN=\" + str(merged.loc[merged.projects == project].shape[0]), fontsize = 40)\n",
    "\n",
    "# plt.legend(title='Folder structure', loc='upper right',\\\n",
    "#            bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_aspect('equal')\n",
    "    ax.legend().remove()\n",
    "    \n",
    "    sns.despine()\n",
    "    ax.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    ax.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.savefig(\"../figures/all_nonproject_folders_tsne_240318.pdf\", bbox_inches = \"tight\", dpi = 300)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73a5f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged['extension'] = merged['path'].str.split('---').str[0].str.split('.').str[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d96339e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.loc[((merged.color== 'black')|(merged.projects.isin(project_folders))\n",
    "             | (merged.projects == \"LSAC\"))].shape[0] - merged.loc[merged.folders == \"Celera\"].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c82a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.loc[~((merged.color== 'black')|(merged.projects.isin(project_folders))\n",
    "             | (merged.projects == \"LSAC\") | (merged.folders.isin(non_project_folders)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2df55924",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.loc[merged.folders.isin(non_project_folders)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8032b87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba97e9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "15117 + 6953"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b0a9c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "15117/22070"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f024eae9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05c30a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create layout\n",
    "layout = [\n",
    "    [0, 1, 2],\n",
    "    [3, 4, '.'],\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplot_mosaic(layout, figsize=(35,35))\n",
    "\n",
    "for ax, extension in zip(axes.values(), merged.extension.unique()):\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    \n",
    "    sns.scatterplot(data = merged.loc[merged.extension != extension], x = 'dim1', y = 'dim2', \\\n",
    "                     color = \"#d3d3d3\",\\\n",
    "                    alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "    sns.scatterplot(data = merged.loc[merged.extension == extension], x = 'dim1', y = 'dim2', \\\n",
    "                color = \"red\",\\\n",
    "                alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "\n",
    "\n",
    "# plt.legend(title='Folder structure', loc='upper right',\\\n",
    "#            bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_aspect('equal')\n",
    "    ax.legend().remove()\n",
    "    size = merged.loc[merged.extension == extension].shape[0]\n",
    "    ax.set_title(f\"{extension}\\nN={size}\", fontsize = 40)\n",
    "    sns.despine()\n",
    "    ax.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    ax.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "634e70e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "1010 + 310 + 20310 + 364 + 76"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1f2d93f",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_documents_with_dates = pd.read_parquet('../cache/pdfs_word_excel_powerpoint_031924_with_dates_for_all.parquet')[['nodeID', 'date']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0baeeaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged['nodeID'] = merged.path.str.split('---').str[-1]\n",
    "merged_with_dates = pd.merge(merged, all_documents_with_dates, on = 'nodeID', how = 'left')\n",
    "merged_with_dates['decade'] = merged_with_dates['date'].apply(lambda x: (x.year//10)*10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7da1c99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fb27b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create layout\n",
    "layout = [\n",
    "    [0, 1],\n",
    "    [2, 3],\n",
    "    [4, '.']\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplot_mosaic(layout, figsize=(40,40))\n",
    "\n",
    "for ax, decade in zip(axes.values(), sorted(merged_with_dates.decade.unique())):\n",
    "    ax.spines['left'].set_linewidth(0)\n",
    "    ax.spines['bottom'].set_linewidth(0)\n",
    "\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    \n",
    "    if pd.isnull(decade):\n",
    "        size = merged_with_dates.loc[merged_with_dates.decade.isna()]\n",
    "        ax.set_title(f\"No date inferred\\nN={size.shape[0]}\", fontsize = 40)\n",
    "        \n",
    "        sns.scatterplot(data = merged_with_dates.loc[merged_with_dates.decade.notna()], x = 'dim1', y = 'dim2', \\\n",
    "                         color = \"#d3d3d3\",\\\n",
    "                        alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged_with_dates.loc[merged_with_dates.decade.isna()], x = 'dim1', y = 'dim2', \\\n",
    "                    color = \"red\",\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "    else:\n",
    "        size = merged_with_dates.loc[merged_with_dates.decade == decade]\n",
    "        ax.set_title(f\"{int(decade)}s\\nN={size.shape[0]}\", fontsize = 40)\n",
    "        \n",
    "        sns.scatterplot(data = merged_with_dates.loc[merged_with_dates.decade != decade], x = 'dim1', y = 'dim2', \\\n",
    "                         color = \"#d3d3d3\",\\\n",
    "                        alpha = 1, ax = ax, linewidth = 0.5, size = 2)\n",
    "        sns.scatterplot(data = merged_with_dates.loc[merged_with_dates.decade == decade], x = 'dim1', y = 'dim2', \\\n",
    "                    color = \"red\",\\\n",
    "                    alpha = 1, ax = ax, linewidth =0.5, size =2)\n",
    "\n",
    "\n",
    "# plt.legend(title='Folder structure', loc='upper right',\\\n",
    "#            bbox_to_anchor=(1.15, 1.05), frameon = False, ncol = 1)\n",
    "    ax.set_ylabel(None)\n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_aspect('equal')\n",
    "    ax.legend().remove()\n",
    "    \n",
    "    \n",
    "    sns.despine()\n",
    "    ax.tick_params(\n",
    "        axis='x',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        bottom=False,      # ticks along the bottom edge are off\n",
    "        top=False,         # ticks along the top edge are off\n",
    "        labelbottom=False) # labels along the bottom edge are off\n",
    "    ax.tick_params(\n",
    "        axis='y',          # changes apply to the x-axis\n",
    "        which='both',      # both major and minor ticks are affected\n",
    "        left=False,      # ticks along the bottom edge are off\n",
    "        right=False)\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "200a6f0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "724 + 4591 + 7555 + 603 + 8593"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "252397df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93241be1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56f87b28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edfcd0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat([pd.DataFrame([0, 1, 2]), pd.DataFrame([0, 1, 2], index = [2, 1, 0])], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d61e5d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96d0879",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
